### PR TITLE
Improve ISO Date Performance for JSON

### PR DESCRIPTION
**Pull request by WillAyd**
benchmarks below 

```sh
       before           after         ratio
     [9c6771c5]       [5c0f5682]
     <master>         <json-index-dates>
-        231±60ms          188±3ms     0.82  io.json.ToJSON.time_iso_format('split', 'df_td_int_ts')
-        496±20ms         218±40ms     0.44  io.json.ToJSON.time_iso_format('index', 'df_int_float_str')
-        486±20ms          207±2ms     0.42  io.json.ToJSON.time_iso_format('columns', 'df_int_float_str')
-         499±9ms          210±1ms     0.42  io.json.ToJSON.time_iso_format('index', 'df_date_idx')
-        503±20ms          206±2ms     0.41  io.json.ToJSON.time_iso_format('columns', 'df_int_floats')
-        515±20ms          210±1ms     0.41  io.json.ToJSON.time_iso_format('index', 'df')
-        524±80ms        209±0.5ms     0.40  io.json.ToJSON.time_iso_format('index', 'df_int_floats')
-        528±10ms          206±2ms     0.39  io.json.ToJSON.time_iso_format('columns', 'df_td_int_ts')
diff --git a/asv_bench/asv.conf.json b/asv_bench/asv.conf.json
index c04bbf53a..897de2f85 100644
--- a/asv_bench/asv.conf.json
+++ b/asv_bench/asv.conf.json
-        546±80ms        209±0.7ms     0.38  io.json.ToJSON.time_iso_format('index', 'df_td_int_ts')
-        568±60ms          208±3ms     0.37  io.json.ToJSON.time_iso_format('columns', 'df_date_idx')
-       598±100ms          206±2ms     0.35  io.json.ToJSON.time_iso_format('columns', 'df')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Note that this mostly improves on DTI which on 0.25.3 can't even be written as ISO format, so I didn't add a whatsnew. Timedelta is the big bottleneck remaining

### CODE REVIEW

**Comment by jbrockmendel** [comment]
CI notwithstanding, this looks like a nice cleanup+speedup

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 1]
```diff
@@ -481,6 +477,19 @@ static char *PyDateTimeToIso(JSOBJ obj, JSONTypeContext *tc, size_t *len) {
     return result;
 }
 
+/* JSON callback */
+static char *PyDateTimeToIsoCallback(JSOBJ obj, JSONTypeContext *tc,
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c]: In another follow up I think I'll move the conversion routines to another file and keep only the core JSON serialization functionality within this ones; makes it a little easier to grok the difference between functions used as callbacks and those used to convert values into various formats

**Comment by jbrockmendel** [comment]
LGTM

File: pandas/tests/io/json/test_pandas.py [hunk 2]
```diff
@@ -854,7 +854,7 @@ def test_date_format_frame(self, date, date_unit):
             json = df.to_json(date_format="iso")
         result = read_json(json)
         expected = df.copy()
-        # expected.index = expected.index.tz_localize("UTC")
+        expected.index = expected.index.tz_localize("UTC")
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: so this fixes a bug?

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 3]
```diff
@@ -1583,54 +1591,36 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, PyObjectEncoder *enc,
...
+            npy_int64 longVal;
+            PyArray_VectorUnaryFunc *castfunc =
+                PyArray_GetCastFunc(PyArray_DescrFromType(type_num), NPY_INT64);
+            if (!castfunc) {
+                PyErr_Format(PyExc_ValueError,
+                             "Cannot cast numpy dtype %d to long",
+                             enc->npyType);
             }
-
+            castfunc(dataptr, &longVal, 1, NULL, NULL);
```
Comment (jreback) [inline pandas/_libs/src/ujson/python/objToJSON.c]: do you need a Py_DECREF anywhere here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 3 above)
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c]: Don't think so - here's a snippet in NumPy where there is no DECREF

https://github.com/numpy/numpy/blob/5ce770ae3de63861c768229573397cadd052f712/numpy/core/src/multiarray/scalarapi.c#L212

And testing locally did segfault when trying

File: pandas/tests/io/json/test_pandas.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_pandas.py]: No this was intentionally commented out here https://github.com/pandas-dev/pandas/pull/27618/files#r311722007 just coming full circle on that now that date alignment is consistent in JSON whether an item appears in the index or in the values

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, might be worth a general whatsnew note for perf for json changes (e.g. list issues that impacted perf that you have done recently)

**Comment by WillAyd** [comment]
Yea no problem; will address in follow up

**Pull request merged**